staging/src/k8s.io/component-base/metrics/prometheus/controllers/metrics.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+var (
+	// Metrics provides access to all controllerreadiness metrics.
+	Metrics = newControllerMetrics()
+)
+
+// ControllerMetrics includes all the metrics of the proxy server.
+type ControllerMetrics struct {
+	controllerInstanceCount *k8smetrics.GaugeVec
+}
+
+// newControllerMetrics create a new ControllerMetrics, configured with default metric names.
+func newControllerMetrics() *ControllerMetrics {
+	controllerInstanceCount := k8smetrics.NewGaugeVec(
+		&k8smetrics.GaugeOpts{
+			Name: "managed_controller_instance_count",
+			Help: "Instances of individual controllers currently running",
+		},
+		[]string{"controller_name", "controller_manager"},
+	)
+	legacyregistry.MustRegister(controllerInstanceCount)
+	return &ControllerMetrics{
+		controllerInstanceCount: controllerInstanceCount,
+	}
+}
+
+// ControllerStarted sets the controllerInstanceCount to 1.
+func (a *ControllerMetrics) ControllerStarted(controllerName string, controllerManager string) {
+	a.controllerInstanceCount.With(k8smetrics.Labels{"controller_name": controllerName, "controller_manager": controllerManager}).Set(float64(1))
+}
+
+// ControllerStopped sets the controllerInstanceCount to 0.
+func (a *ControllerMetrics) ControllerStopped(controllerName string, controllerManager string) {
+	a.controllerInstanceCount.With(k8smetrics.Labels{"controller_name": controllerName, "controller_manager": controllerManager}).Set(float64(0))
```

Comment:
Is there already a guarantee that each controller can only possibly be started once? I would imagine this metric would be very misleading if somehow we have a zombie controller running. (In which case I'd prefer this do an increment on start and decrement on stop).

Reply:
So this is designed to detect zombie controllers. Since it'll be a new metric per-controller, we would see two instances of a metric with the "1" state, and total them to determine if there were too many (or too few) controllers running

Reply:
Set(0) and Set(1) works for making sure we don't have the same controller elected across N cluster-manager instances.

If we don't think it's possible to every have >1 instance of the same controller running in a controller manager, Set(0) and Set(1) are safe, but I don't know enough about the controller manager to know if that's a safe assumption. If it is, a comment explaining would help here, because it's pretty surprising looking code to the naive reader.

Reply:
@cheftako can you comment to this? I have updated with an in-line code comment

Reply:
So I think there are two possible use cases for multiple instances of a controller. This code is clearly aimed at resolving multiple process (multiple manager) related issues. For that I think its fine. Joe is bringing up the idea that we could have similar problems within A process (A manager). Without the context of where these utilities are being called it is hard to understand what you are solving for. An example of Joe's concern would be if the controller-name to controller-init map had an error in it which caused different names to resolve to one controller-init. A related issue would be if we ever tried to have controller recovery. I think this is an improvement for now but I think it would be good to have a bug to follow up in the way that Joe is suggesting.

Reply:
I'm can live with follow up. For the follow up, I'd be fine with either:

- option A: We ensure that the existing code has preconditions in place that prevent two controllers with the same identity from being started at the same time, AND that there is a test showing this
- option B: We acknowledge that it is possible to for two controllers with the same identity from being started at the same time, we use inc/dec here and add a test showing the behavior is as expected

I like option A the best, for reasons that are hopefully obvious. And the precondition already be in place? If so I would really like a comment in this code stating that so it's clear to future readers of the code why the set(0)/set(1) is safe.

Reply:
Wouldn't the following be more appropriate here:
```suggestion
	a.controllerInstanceCount.With(k8smetrics.Labels{"controller_name": controllerName, "controller_manager": controllerManager}).Add(float64(-1))
```
In case we have multiple instances of a controller, stopping one should result in the current count of instances being decreased by 1 no?

Reply:
> Wouldn't the following be more appropriate

That was my original question. It sounds like there is suppose to be an invariant preventing it and there are downsides both ways (ways to miscount either way depending on what goes wrong).